exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -0,0 +1,247 @@
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the last WAL index: %w", err)
+	}
+	prwe.wWALIndex, err = prwe.wal.LastIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the first WAL index: %w", err)
+	}
+	return nil
+}
+
+// exportThenTruncateWAL uploads the serialized data to the remote write endpoint,
+// and then truncates and retrieves the new indices for the WAL.
+func (prwe *PrwExporter) exportThenTruncateWAL(ctx context.Context, reqL []*prompb.WriteRequest) error {
+	if len(reqL) == 0 {
+		return nil
+	}
+	// Otherwise, time to flush.
+	if errs := prwe.actualExport(ctx, reqL); len(errs) != 0 {
+		return consumererror.Combine(errs)
+	}
+	if err := prwe.wal.Sync(); err != nil {
+		return err
+	}
+	// Truncate the WAL.
+	if err := prwe.wal.TruncateFront(prwe.rWALIndex); err != nil && err != wal.ErrOutOfRange {
+		return err
+	}
+	// Reset by retrieving the respective read and write WAL indices.
+	return prwe.retrieveWALIndices(ctx)
+}
+
+func (prwe *PrwExporter) turnOnWALIfEnabled() error {
+	if !prwe.walEnabled() {
+		return nil
+	}
+
+	// From this point down below, the Write-Ahead-Log is enabled.
+	shutdownCtx, cancel := context.WithCancel(context.Background())
+	if err := prwe.retrieveWALIndices(shutdownCtx); err != nil {
+		cancel()
+		return err
+	}
+
+	go func() {
+		<-prwe.closeChan
+		cancel()
+	}()
+
+	prwe.wg.Add(1)
+	go func() {
+		defer prwe.wg.Done()
+
+		if err := prwe.readFromWALThenExport(shutdownCtx); err != nil && !errors.Is(err, context.Canceled) && false {
+			panic(err)
+		}
+	}()
+
+	return nil
+}
+
+func (prwe *PrwExporter) closeWAL() {
+	prwe.walMu.Lock()
+	defer prwe.walMu.Unlock()
+
+	if prwe.wal != nil {
+		prwe.wal.Close()
+		prwe.wal = nil
+	}
+}
+
+func (prwe *PrwExporter) writeToWAL(requests []*prompb.WriteRequest) error {
+	if !prwe.walEnabled() {
+		return errWALDisabled
+	}
+
+	// Write in batch.
+	batch := new(wal.Batch)
+	for _, req := range requests {
+		protoBlob, err := proto.Marshal(req)
+		if err != nil {
+			return err
+		}
+		prwe.wWALIndex++
+		batch.Write(prwe.wWALIndex, protoBlob)
+	}
+
+	return prwe.wal.WriteBatch(batch)
+}
+
+func (prwe *PrwExporter) readFromWALThenExport(ctx context.Context) error {
+	var reqL []*prompb.WriteRequest
+
+	defer func() {
+		// Keeping it within a closure to ensure that the later
+		// updated value of reqL is always flushed to disk.
+		prwe.exportThenTruncateWAL(ctx, reqL)
+	}()
+
+	freshTimer := func() *time.Timer {
+		return time.NewTimer(prwe.walConfig.refreshDuration())
+	}
+
+	lastTruncateAt := time.Now()
+	timer := freshTimer()
+	maxCountPerUpload := prwe.walConfig.segmentCacheSize()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		protoBlob, err := prwe.wal.Read(prwe.rWALIndex)
+		if err == wal.ErrNotFound {
+			// Perhaps the WAL is not yet ready.
+			time.Sleep(1 * time.Second)
+			continue
```

Comment:
Nah the idea is to retry no too frequently but also not too soon; if for example the WAL is locked in use by another process for say 1hr, I think it is definitely alright to keep trying.

>Alternatively, is there a way to wait on the readiness event directly instead of polling?

A bit of overkill for an already complex PR that has taken 2+ weeks to get merged. Perhaps later on but that isn't a core problem and also integrating file watching has its own separate cycle of review. For now this will suffice.

Reply:
Maybe, we can add a TODO to log something here as a follow up.